### PR TITLE
Add "Write components that are easy to test"

### LINKED ISSUE #34: feat: Add "Write Vue code that's easy to test" section
I was thinking of adding a section somewhere with several tips on writing testable code:

* Extract requests into services to simplify mocking.
* Extract complex methods and test them in isolation.
* Do not test implementation details (_with examples_).
* Build small things to have simpler test scenarios.
* Write tests before writing production code (_the ultimate technique to write testable code_ 😆).

Also, we could point to some smells:

* Large "setup/arrange" blocks on each test.
* Breaking tests on a refactor (_from my POV this is critical but I'm not 100% sure it would fit in this section_).

Sounds cool? More ideas?

disclaimer: There's a lot of content on the internet about these topics. I think this section should be focused on Vue stuff with Vue examples, and then maybe links to well-established content.

### PR DESCRIPTION
**Pull request by afontcu**
Closes #34

### CODE REVIEW

File: src/guide/write-easy-test.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+### Outputs
+
+**DOM elements**: anything the component renders to the DOM.
+
+**events**: emitted (through `$emit`) events.
+
+**side effects**: any other *observable* side effect, such as `console.log`, cookie creation, API calls…
+
+## Everything else are implementation details
```
Comment (lmiller1990) [inline src/guide/write-easy-test.md:29]: Everything else *is* implementation details

File: src/guide/write-easy-test.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+Notice how this list does not include elements such as internal methods, or intermediate states or data.
+
+The rule of thumb is that **a test should not break on a refactor**, that is, when we change its internal implementation without changing its public API. If that happens, it might mean the test relies on implementation details.
```
Comment (lmiller1990) [inline src/guide/write-easy-test.md:33]: This is probably the best way to explain it. It would be great to include a test that does everything wrong, and show how a refactor breaks it, then show how writing it based in input/outputs makes it resistant to refactors.

File: src/guide/write-easy-test.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+## Build smaller components
+
+If a component does less, then it will be easier to test.
```
Comment (lmiller1990) [inline src/guide/write-easy-test.md:38]: this is an interesting one. sometimes this can go too far, and people write tiny components and unit test those, which is kind of like testing an implementation detail (not really, but you can imagine if someone had 10 components for a form, and tested each one individually, that would not be as useful as 1 test on the form itself).

this is just my opinion though - I don't have a test file for every component, but I try to have tests that cover all of the components (indirectly is fine, like a form). not sure how opinionated we should be here, though.

File: src/guide/write-easy-test.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+### Extract API calls
+
+:::tip
+ Check out the [Making HTTP requests](../guide/http-requests.md) guide if you are unfamiliar with testing API calls.
+:::
+
+Usually, you will perform several HTTP requests throughout your application. From a testing perspective, HTTP requests provide inputs to the component, and a component can also send HTTP requests.
```
Comment (lmiller1990) [inline src/guide/write-easy-test.md:47]: great, i love this

File: src/guide/write-easy-test.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+### Extract complex methods
+
+Sometimes a component might feature a complex method, with heavy calculations or several dependencies.
+
+The suggestion here is to extract this method and import it to the component. This way, you can test the method in isolation (if necessary), and you can mock the import when testing your component.
```
Comment (lmiller1990) [inline src/guide/write-easy-test.md:54]: it would be super great to include an example of this here too, if you have time.

**Review by lmiller1990 — COMMENTED** [review]
Various comments, mostly just food for thought - happy to let you own this one and do what like with it!

File: src/guide/write-easy-test.md (hunk 2 above)
Comment (afontcu) [reply src/guide/write-easy-test.md:33]: great idea! I'll look into it to come up with a simple but meaningful example

File: src/guide/write-easy-test.md (hunk 5 above)
Comment (afontcu) [reply src/guide/write-easy-test.md:54]: sure!

**Review by hc0503 — APPROVED** [review]

File: src/guide/write-easy-test.md (hunk 3 above)
Comment (afontcu) [reply src/guide/write-easy-test.md]: I rephrase it to make it read as a suggestion:

"A general rule of thumb is that if a component does less, then it will be easier to test."
